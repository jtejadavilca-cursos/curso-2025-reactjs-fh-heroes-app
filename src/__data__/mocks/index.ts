export { mockHero } from "./mock-hero";
export { mockSummaryData } from "./mock-summary-data";
