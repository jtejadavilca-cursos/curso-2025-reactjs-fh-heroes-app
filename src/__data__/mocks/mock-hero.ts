import type { Hero } from "@/heroes/types/hero.interface";
export const mockHero = {
    id: 123,
    name: "My Test hero",
} as unknown as Hero;
