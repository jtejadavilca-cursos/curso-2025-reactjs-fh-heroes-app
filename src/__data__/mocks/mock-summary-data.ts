import type { SummaryInformationResponse } from "@/heroes/types/summary-information.response";

export const mockSummaryData: SummaryInformationResponse = {
    totalHeroes: 25,
    strongestHero: {
        id: "1",
        name: "Clark Kent",
        slug: "clark-kent",
        alias: "Superman",
        powers: [
            "Súper fuerza",
            "Vuelo",
            "Visión de calor",
            "Visión de rayos X",
            "Invulnerabilidad",
            "Súper velocidad",
        ],
        description: "El Último Hijo de Krypton, protector de la Tierra y símbolo de esperanza para toda la humanidad.",
        strength: 10,
        intelligence: 8,
        speed: 9,
        durability: 10,
        team: "Liga de la Justicia",
        image: "1.jpeg",
        firstAppearance: "1938",
        status: "Active",
        category: "Hero",
        universe: "DC",
    },
    smartestHero: {
        id: "2",
        name: "Bruce Wayne",
        slug: "bruce-wayne",
        alias: "Batman",
        powers: ["Artes marciales", "Habilidades de detective", "Tecnología avanzada", "Sigilo", "Genio táctico"],
        description:
            "El Caballero Oscuro de Ciudad Gótica, que utiliza el miedo como arma contra el crimen y la corrupción.",
        strength: 6,
        intelligence: 10,
        speed: 6,
        durability: 7,
        team: "Liga de la Justicia",
        image: "2.jpeg",
        firstAppearance: "1939",
        status: "Active",
        category: "Hero",
        universe: "DC",
    },
    heroCount: 18,
    villainCount: 7,
};
