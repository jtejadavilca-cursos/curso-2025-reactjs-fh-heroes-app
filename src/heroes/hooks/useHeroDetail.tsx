import { useQuery } from "@tanstack/react-query";
import { getHeroeByIdOsSlugAction } from "../actions/get-hero.action";

export const useHeroDetail = (idSlug: string) => {
    const { data: superheroData } = useQuery({
        queryKey: ["heroes", "idSlug", idSlug],
        queryFn: () => getHeroeByIdOsSlugAction(idSlug),
        staleTime: 1000 * 60 * 5, // 5 minutos
    });

    const totalPower = superheroData
        ? superheroData.strength + superheroData.intelligence + superheroData.speed + superheroData.durability
        : 0;
    const averagePower = Math.round((totalPower / 4) * 10);

    return {
        superheroData,
        totalPower,
        averagePower,
    };
};
