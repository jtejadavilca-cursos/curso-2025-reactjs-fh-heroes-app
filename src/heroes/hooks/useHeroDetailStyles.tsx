// const mapStatusColors = new Map<HeroStatus, string>();
const mapStatusColors = new Map<string, string>();
mapStatusColors.set("activo", "bg-green-500");
mapStatusColors.set("inactivo", "bg-gray-500");
mapStatusColors.set("retirado", "bg-blue-500");

// const mapCategoryColors = new Map<HeroCategory, string>();
const mapCategoryColors = new Map<string, string>();
mapCategoryColors.set("héroe", "bg-blue-500");
mapCategoryColors.set("villano", "bg-red-500");
mapCategoryColors.set("antihéroe", "bg-purple-500");

const getStatusColor = (status: string) => mapStatusColors.get(status.toLowerCase()) ?? "bg-gray-500";
const getCategoryColor = (category: string) => mapCategoryColors.get(category.toLowerCase()) ?? "bg-gray-500";

export const useHeroDetailStyles = () => {
    return {
        getStatusColor,
        getCategoryColor,
    };
};
