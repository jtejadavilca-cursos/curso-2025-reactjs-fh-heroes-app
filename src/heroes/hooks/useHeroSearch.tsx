import { useQuery } from "@tanstack/react-query";
import type { SearchProps } from "../actions/search-heroes.action";
import { getSearchAction } from "../actions/search-heroes.action";

export const useHeroSearch = (filters: SearchProps) => {
    const cleanedFilters = Object.fromEntries(Object.entries(filters).filter(([_, value]) => !!value)) as SearchProps;

    return useQuery({
        queryKey: ["search-heroes", { ...cleanedFilters }],
        queryFn: () => getSearchAction(cleanedFilters),
        staleTime: 1000 * 60 * 5,
    });
};
