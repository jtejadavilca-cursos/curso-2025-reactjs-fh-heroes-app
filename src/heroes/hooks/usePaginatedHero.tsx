import { useQuery } from "@tanstack/react-query";
import { getHeroesByPageAction } from "../actions/get-heroes-by-page.action";

export const usePaginatedHero = (page: number, limit: number, category: string = "all") => {
    const sanitizedPage = isNaN(page) ? 1 : +page;
    const sanitizedLimit = isNaN(limit) ? 6 : +limit;

    return useQuery({
        queryKey: ["heroes", { page: sanitizedPage, limit: sanitizedLimit, category }],
        queryFn: () => getHeroesByPageAction(sanitizedPage, sanitizedLimit, category),
        staleTime: 1000 * 60 * 5, // 5 minutos
    });
};
