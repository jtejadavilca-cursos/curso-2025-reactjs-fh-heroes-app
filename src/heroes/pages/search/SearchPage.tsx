import { useSearchParams } from "react-router";
import { CustomJumbotron } from "@/components/custom/CustomJumbotron";
import { HeroStats } from "@/heroes/components/HeroStats";
import { useHeroSearch } from "@/heroes/hooks/useHeroSearch";
import { HeroGrid } from "@/heroes/components/HeroGrid";
import { SearchControls } from "./ui/SearchControls";

export const SearchPage = () => {
    const [searchParams] = useSearchParams();

    const { data: searchResult } = useHeroSearch({
        name: searchParams.get("name") ?? undefined,
        strength: searchParams.get("strength") ? Number(searchParams.get("strength")) : undefined,
    });

    return (
        <>
            {/* Header */}
            <CustomJumbotron
                title="Búsqueda de Superheroes!"
                description="Descubre, explora, y administra tus superhéroes y villanos favoritos!"
            />

            {/* Stats Dashboard */}
            <HeroStats />
            {/* Controls */}
            <SearchControls />

            {/* Search Result */}
            <HeroGrid heroes={searchResult ?? []} />
        </>
    );
};

export default SearchPage;
