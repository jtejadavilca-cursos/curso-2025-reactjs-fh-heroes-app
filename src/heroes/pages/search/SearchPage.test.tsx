import { beforeEach, describe, expect, test, vi } from "vitest";
import SearchPage from "./SearchPage";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { useHeroSearch } from "@/heroes/hooks/useHeroSearch";
import type { Hero } from "@/heroes/types/hero.interface";
import { mockHeros } from "@/__data__/mocks/mock-hero";

vi.mock("@/heroes/components/HeroGrid", () => ({
    HeroGrid: ({ heroes }: { heroes: Hero[] }) => (
        <div data-testid="mocked-herogrid">Mocked HeroGrid - Total heroes: {heroes.length}</div>
    ),
}));

vi.mock("./ui/SearchControls", () => ({
    SearchControls: () => <div data-testid="mocked-search-control">Mocked SearchControls</div>,
}));

vi.mock("@/heroes/hooks/useHeroSearch");
const mockUseHeroSearch = vi.mocked(useHeroSearch);

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            retry: false,
        },
    },
}); //

const renderHomePage = (
    initialEntries?: string[], //
    searchResult: Hero[] = []
) => {
    mockUseHeroSearch.mockReturnValue({
        data: searchResult,
    } as unknown as ReturnType<typeof useHeroSearch>);

    return render(
        <MemoryRouter initialEntries={initialEntries}>
            {/* <FavoriteHeroProvider> */}
            <QueryClientProvider client={queryClient}>
                <SearchPage />
            </QueryClientProvider>
            {/* </FavoriteHeroProvider> */}
        </MemoryRouter>
    );
};

describe("Testing SearchPage.test", () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    test("should render SearchPage with default values", () => {
        const { container } = renderHomePage(["/"], []);

        expect(mockUseHeroSearch).toHaveBeenCalledWith({
            name: undefined,
            strength: undefined,
        });

        expect(container).toMatchSnapshot();
    });

    test.each([
        {
            searchQuery: "",
            searchParams: {},
        },
        {
            searchQuery: "/search?name=superman",
            searchParams: { name: "superman" },
        },
        {
            searchQuery: "/search?strength=5",
            searchParams: { strength: 5 },
        },
        {
            searchQuery: "/search?name=superman&strength=5",
            searchParams: { name: "superman", strength: 5 },
        },
    ])(
        "should call search action with parameters $searchParams for queries $searchQuery",
        ({ searchQuery, searchParams }) => {
            renderHomePage([searchQuery], []);
            expect(mockUseHeroSearch).toHaveBeenCalledWith(searchParams);
        }
    );

    test("should render HeroGrid with search results", async () => {
        //
        renderHomePage([""], mockHeros);

        expect(screen.getByTestId("mocked-herogrid")).toBeDefined();
        expect(screen.getByTestId("mocked-herogrid").innerHTML).toContain("Mocked HeroGrid - Total heroes: 5");

        screen.debug();
    });
});
