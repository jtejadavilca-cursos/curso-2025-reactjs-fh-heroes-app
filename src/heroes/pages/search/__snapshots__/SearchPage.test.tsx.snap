// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing SearchPage.test > should render SearchPage with default values 1`] = `
<div>
  <div
    class="text-center mb-8"
  >
    <h1
      class="text-5xl font-bold bg-gradient-to-r from-blue-600 to-purple-600 bg-clip-text text-transparent mb-4"
    >
      Búsqueda de Superheroes!
    </h1>
    <p
      class="text-gray-600 text-lg"
    >
      Descubre, explora, y administra tus superhéroes y villanos favoritos!
    </p>
  </div>
  <div>
    Loading...
  </div>
  <div
    data-testid="mocked-search-control"
  >
    Mocked SearchControls
  </div>
  <div
    data-testid="mocked-herogrid"
  >
    Mocked HeroGrid - Total heroes: 
    0
  </div>
</div>
`;
